Spigot-Server-Patches/0246-Port-Sponge-Async-lighting-patch-to-Paper.patch

```diff
@@ -0,0 +1,1221 @@
+             return super.a(i, j, k).h();
+         }
+diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
+index e818f7e3e..518411f06 100644
+--- a/src/main/java/net/minecraft/server/Chunk.java
++++ b/src/main/java/net/minecraft/server/Chunk.java
+@@ -26,26 +26,26 @@ public class Chunk {
+     private final ChunkSection[] sections;
+     private final byte[] g;
+     private final int[] h;
+-    private final boolean[] i;
++    private final boolean[] i; public boolean[] getUpdateSkylightColumns() { return i; } // Paper - OBFHELPER
+     private boolean j;
+     public final World world;
+     public final int[] heightMap;
+     public Long scheduledForUnload; // Paper - delay chunk unloads
+     public final int locX;
+     public final int locZ;
+-    private boolean m;
++    private boolean m; public boolean isGapLightingUpdate() { return m; } public void setGapLightingUpdate(boolean value) { m = value; } // Paper - OBFHELPER
+     public final Map<BlockPosition, TileEntity> tileEntities;
+     public final List<Entity>[] entitySlices; // Spigot
+     private boolean done;
+     private boolean lit;
+-    private boolean r;
+-    private boolean s;
++    private boolean r; public boolean isTicked() { return r; } public void setTicked(boolean value) { r = value; } // Paper - OBFHELPER
++    private boolean s; public boolean isDirty() { return s; } public void setDirty(boolean value) { s = value; } // Paper - OBFHELPER
+     private boolean t;
+     private long lastSaved;
+-    private int v;
++    private int v; public int getHeightMapMinimum() { return v; } public void setHeightMapMinimum(int value) { v = value; } // Paper - OBFHELPER
+     private long w;
+     private int x;
+-    private final ConcurrentLinkedQueue<BlockPosition> y;
++    private final ConcurrentLinkedQueue<BlockPosition> y; public ConcurrentLinkedQueue<BlockPosition> getTileEntityPosQueue() { return y; } // Paper - OBFHELPER
+     public boolean d; public void setShouldUnload(boolean unload) { this.d = unload; } public boolean isUnloading() { return d; } // Paper - OBFHELPER
+     protected gnu.trove.map.hash.TObjectIntHashMap<Class> entityCount = new gnu.trove.map.hash.TObjectIntHashMap<Class>(); // Spigot
+ 
+@@ -123,6 +123,7 @@ public class Chunk {
+     // CraftBukkit end
+ 
+     public Chunk(World world, int i, int j) {
++        lightExecutorService = world.getLightingExecutor(); // Paper - asynclighting
+         this.sections = new ChunkSection[16];
+         this.g = new byte[256];
+         this.h = new int[256];
+@@ -151,6 +152,18 @@ public class Chunk {
+     public boolean mustSave;
+     // CraftBukkit end
+ 
++    // Paper start - asynclighting
++    // Keeps track of block positions in this chunk currently queued for sky light update
++    private java.util.concurrent.CopyOnWriteArrayList<Short> queuedSkyLightingUpdates = new java.util.concurrent.CopyOnWriteArrayList<>();
```

Comment:
Can we use a ConcurrentLinkedQueue for all of these?

COWAL imo should not be used for things you expect lots of changes to.